File: codewars/4-kyu/infix/a.py
from collections import deque


def to_postfix(infix: str):
  """Convert infix to postfix"""

  de = deque([])
  val = ""
  prec = {
      '+': 1,
      '-': 1,
      '*': 2,
      '/': 2,
      '^': 3,
      '(': -1
  }

  for ch in infix:

    if ch.isnumeric():
      val += ch

    elif ch == "(":
      de.append(ch)

    elif ch == ")":
      while de[-1] != "(":
        val += de.pop()

      de.pop()

    else:

      while (len(de) != 0 and prec[ch] <= prec[de[-1]]):
        val += de.pop()

      de.append(ch)

  while len(de) != 0:
    val += de.pop()

  return val


if __name__ == '__main__':
  print(to_postfix("1^2^3"))

File: codewars/4-kyu/infix/a.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int prec(char ch)
{
  if (ch == '+' || ch == '-')
  {
    return 1;
  }
  else if (ch == '*' || ch == '/')
  {
    return 2;
  }
  else if (ch == '^')
  {
    return 3;
  }

  // It will only return -1 for '('
  return -1;
}

std::string to_postfix(std::string infix)
{
  stack<char> st;
  string val = "";

  for (auto &ch : infix)
  {
    if (isdigit(ch))
    {
      val += ch;
    }

    else if (ch == '(')
    {
      st.push(ch);
    }

    else if (ch == ')')
    {
      while (st.top() != '(')
      {
        val += st.top();
        st.pop();
      }
      st.pop();
    }

    else
    {
      while (!st.empty() && (prec(ch) <= prec(st.top())))
      {
        val += st.top();
        st.pop();
      }

      st.push(ch);
    }
  }

  while (!st.empty())
  {
    val += st.top();
    st.pop();
  }

  return val;
}

int main()
{
  string s = "1+2+3";

  cout << to_postfix(s) << endl;
}